File: ProblemSet/509.fibonacci-number/509.fibonacci-number 记忆化迭代.java
/*
 * @lc app=leetcode id=509 lang=java
 *
 * [509] Fibonacci Number
 */

// @lc code=start
class Solution {
    public int fib(int N) {
        if (N <= 1) return N;
        int pre1 = 0, pre2 = 1, cur = 1;
        for (int i = 2; i <= N; ++i) {
            cur = pre1 + pre2;
            pre1 = pre2;
            pre2 = cur;
        }
        return cur;
    }
}
// @lc code=end

File: ProblemSet/509.fibonacci-number/python/509.fibonacci-number 记忆化迭代.py
#
# @lc app=leetcode id=509 lang=python3
#
# [509] Fibonacci Number
#

# @lc code=start
class Solution:
    def fib(self, n: int) -> int:
        if n <= 1:
            return n
        pre1, pre2 = 0, 1
        for _ in range(2, n + 1):
            pre1, pre2 = pre2, pre1 + pre2
        return pre2
        
# @lc code=end

